src/stash/src/upgrade/json_to_proto.rs

```diff
@@ -0,0 +1,1313 @@
+const COLLECTION_CLUSTER_INTROSPECTION_SOURCE_INDEX_NAME: &str =
+    "compute_introspection_source_index";
+const COLLECTION_CLUSTER_REPLICAS_NAME: &str = "compute_replicas";
+const COLLECTION_DATABASE_NAME: &str = "database";
+const COLLECTION_SCHEMA_NAME: &str = "schema";
+const COLLECTION_ITEM_NAME: &str = "item";
+const COLLECTION_ROLE_NAME: &str = "role";
+const COLLECTION_TIMESTAMP_NAME: &str = "timestamp";
+const COLLECTION_SYSTEM_CONFIGURATION_NAME: &str = "system_configuration";
+const COLLECTION_AUDIT_LOG_NAME: &str = "audit_log";
+const COLLECTION_STORAGE_USAGE_NAME: &str = "storage_usage";
+
+const COLLECTION_STORAGE_METADATA: &str = "storage-collection-metadata";
+const COLLECTION_STORAGE_METADATA_EXPORT: &str = "storage-export-metadata-u64";
+const COLLECTION_STORAGE_SHARD_FINALIZATION: &str = "storage-shards-to-finalize";
+
+const INSERT_BATCH_MAX_SIZE: usize = 4 * 1024 * 1024;
+
+const DATA_PROTO_DROP_TABLE_QUERY: &str = "DROP TABLE IF EXISTS data_proto";
+const DATA_PROTO_TABLE_QUERY: &str = "
+CREATE TABLE data_proto (
+    collection_id bigint NOT NULL REFERENCES collections (collection_id),
+    key bytea NOT NULL,
+    value bytea NOT NULL,
+    time bigint NOT NULL,
+    diff bigint NOT NULL
+);
+";
+
+/// A __special__ migration that migrates all of the types in the `data` table from JSON to
+/// protobuf.
+pub async fn migrate_json_to_proto(client: &mut Client) -> Result<(), anyhow::Error> {
+    // Create a new table for the protobufs.
+    //
+    // Note: Cockroach does not support selecting or inserting into a table that you created in the
+    // same transaction _if_ that table references another, e.g. how data_proto references the
+    // collections table. Because of this, we create the data_proto table outside of the transaction.
+    client
+        .execute(DATA_PROTO_DROP_TABLE_QUERY, &[])
+        .await
+        .context("dropping data_proto, if exists")?;
+    client
+        .execute(DATA_PROTO_TABLE_QUERY, &[])
+        .await
+        .context("creating data_proto")?;
+
+    let tx = client
+        .build_transaction()
+        .start()
+        .await
+        .context("staring transaction")?;
+
+    // Get all of our current collections.
+    let rows = tx
+        .query("SELECT name, collection_id FROM collections", &[])
+        .await
+        .context("get all collections")?;
+    let collections: BTreeMap<i64, String> = rows
+        .into_iter()
+        .map(|row| (row.get("collection_id"), row.get("name")))
+        .collect();
+
+    // Migrate all of the collections.
+    migrate_collections(&tx, collections.clone())
+        .await
+        .context("migrating collections")?;
+
+    // Check if the Stash is valid.
+    validate_collections(&tx, collections)
+        .await
+        .context("validating Stash")?;
+
+    // Commit the migrated protos, but don't yet rename data_proto to data.
+    tx.commit().await.context("commiting protos")?;
```

Comment:
This `?` could also fail in interesting ways and needs to get to sentry if it's not already handled by the upstream caller.